### PR TITLE
Use sample_weight in the metric calculations in model validations and cross validation

### PR DESCRIPTION
**Pull request by ryan-deak-zefr**
# *DO NOT MERGE WITH PR WITHOUT ASKING!*

# Purpose

To correct metric calculations in model validation (and cross validation) by attempting to use `sample_weight` in the metric calculations when `sample_weight` are supplied in `fit_params`.

Adds sample weighting for model validation methods like cross validation inside `sklearn/model_selection/_validation.py` and `sklearn/model_selection/_search.py`.

#### Reference Issues/PRs

Fixes #4632.  See also #10806.

#### What does this implement/fix? Explain your changes.

This change always uses the same `sample_weight` used in training to weight the metric calculations in `_fit_and_score` and functions called by `_fit_and_score`, so it should be thought of changing the default behavior of things like cross validation.  This is similar to PR #10806 but doesn't allow separate test weights to be specified.  It changes `_validation.py` and `_search.py`.

None of the CV search classes (`GridSearchCV`, etc) are changed.  This operates at the function level.

#### Any other comments?

This PR provides some simple test showing how unweighted metrics have very unfavorable consequences when models are trained with importance weights but not scored with the same weights.  It is shown in the test that the current omission of `sample_weight` in metrics calculations considerably overestimates the accuracy.

### CODE REVIEW

File: sklearn/model_selection/_search.py [hunk 1]
```diff
@@ -788,9 +834,29 @@ def _store(key_name, array, weights=None, splits=False, rank=False):
...
 
-        # NOTE test_sample counts (weights) remain the same for all candidates
-        test_sample_counts = np.array(test_sample_counts[:n_splits],
-                                      dtype=np.int)
+        # training train_sample_weight_sums needs to be done first because
+        # test_sample_counts overrides itself in the IF statement below with
+        # test_sample_weight_sums.
+        if self.return_train_score:
+            if train_sample_weight_sums is None:
+                samples = int(np.sum(test_sample_counts[:n_splits]))
```
Comment (ryan-deak-zefr) [inline sklearn/model_selection/_search.py:844]: These two lines are the hairiest of all.  I'm not 100% certain about whether this is the right way to do it but unfortunately, I don't see a better way because the total dataset size isn't maintained anywhere inside the function.  Any thoughts?

File: sklearn/model_selection/_search.py [hunk 2]
```diff
@@ -666,6 +696,15 @@ def evaluate_candidates(candidate_params):
                                in product(candidate_params,
                                           cv.split(X, y, groups)))
 
+                out = list(out)
+                if 0 < len(out):
+                    out, train_wts, test_wts = zip(*out)
+                else:
+                    out, train_wts, test_wts = ([], [], [])
+
+                train_wts = collape_nones(train_wts)
```
Comment (ryan-deak-zefr) [inline sklearn/model_selection/_search.py]: Turns lists of `None` into a `None`.

File: sklearn/model_selection/_search.py [hunk 3]
```diff
@@ -666,6 +696,15 @@ def evaluate_candidates(candidate_params):
                                in product(candidate_params,
                                           cv.split(X, y, groups)))
 
+                out = list(out)
+                if 0 < len(out):
+                    out, train_wts, test_wts = zip(*out)
+                else:
+                    out, train_wts, test_wts = ([], [], [])
+
+                train_wts = collapse_nones(train_wts)
```
Comment (ryan-deak-zefr) [inline sklearn/model_selection/_search.py:705]: Turns a list of `None` into a `None`.

File: sklearn/model_selection/_search.py [hunk 4]
```diff
@@ -819,8 +885,11 @@ def _store(key_name, array, weights=None, splits=False, rank=False):
                    splits=True, rank=True,
                    weights=test_sample_counts if iid else None)
             if self.return_train_score:
+                # _store('train_%s' % scorer_name, train_scores[scorer_name],
+                #        splits=True)
                 _store('train_%s' % scorer_name, train_scores[scorer_name],
-                       splits=True)
+                       splits=True,
+                       weights=train_sample_counts if iid else None)
```
Comment (ryan-deak-zefr) [inline sklearn/model_selection/_search.py:891]: I'm not sure about this `iid` parameter.  I copied it from the `test_%s` case a few lines above.  I feel like you want it more when you don't have *IID*, so this is a little weird to me, but that's how it is above.

**Review by ryan-deak-zefr — COMMENTED** [review]
Put a few places that I am concerned about.

File: sklearn/model_selection/_search.py [hunk 5]
```diff
@@ -788,9 +834,29 @@ def _store(key_name, array, weights=None, splits=False, rank=False):
...
-        test_sample_counts = np.array(test_sample_counts[:n_splits],
-                                      dtype=np.int)
+        # training train_sample_weight_sums needs to be done first because
+        # test_sample_counts overrides itself in the IF statement below with
+        # test_sample_weight_sums.
+        if self.return_train_score:
+            if train_sample_weight_sums is None:
+                samples = int(np.sum(test_sample_counts[:n_splits]))
+                train_sample_counts = \
+                    np.array(test_sample_counts[:n_splits], dtype=np.int) - \
```
Comment (zachary-mcpher) [inline sklearn/model_selection/_search.py]: Could you explain why this difference is necessary?

Comment (ryan-deak-zefr) [reply sklearn/model_selection/_search.py]: @zachary-mcpherson-zefr: Good catch!  The arguments are backward.  Will fix.  In the meantime, here's the reasoning (assuming the code was correct):

Say that you have the following three folds for cross validation:

* **fold 1**: *50* examples
* **fold 2**: *51* examples
* **fold 3**: *49* examples

So it typically works by iterating over the test fold.  So let's say that we are on the second loop iteration and therefore, **fold 2** is the test fold.  That means that folds 1 and 3 are the training folds.  So, there should be *99* training examples for test **fold 2**.

`samples` should represent the total number of rows of data in the entire dataset.  Then, to get the number of training rows, we can subtract the number of testing rows from the total to get the number of training rows associated with the test fold.  So, back to our example, if the test fold is **fold 2**, we have `samples = 150` and **fold 2** is *51*, then `150 - 51 == 99`.  Make sense?

File: sklearn/model_selection/_search.py (hunk 1 above)
Comment (vecchp) [reply sklearn/model_selection/_search.py:844]: Can we not store the total data set size in some way and pass it in?

File: sklearn/model_selection/_search.py [hunk 6]
```diff
@@ -648,6 +648,35 @@ def fit(self, X, y=None, groups=None, **fit_params):
             all_candidate_params = []
             all_out = []
 
+            def is_none(x):
+                return x is None
+
+            def collape_nones(xs):
```
Comment (vecchp) [inline sklearn/model_selection/_search.py]: Did you mean collapse nones?

File: sklearn/model_selection/_search.py [hunk 7]
```diff
@@ -648,6 +648,35 @@ def fit(self, X, y=None, groups=None, **fit_params):
...
+                return None if xs is None or any(map(is_none, xs)) else xs
```
Comment (vecchp) [inline sklearn/model_selection/_search.py:655]: Perhaps the function can be written like with a generator instead of a map?

```python
def collapse_nones_g(xs):
    if xs is None or any(x is None for x in xs):
        return None
    return xs
```

**Review by vecchp — CHANGES_REQUESTED** [review]

File: sklearn/model_selection/_search.py (hunk 5 above)
Comment (zachary-mcpher) [reply sklearn/model_selection/_search.py]: Yeah that makes sense, thanks!

File: sklearn/model_selection/_search.py (hunk 7 above)
Comment (ryan-deak-zefr) [reply sklearn/model_selection/_search.py:655]: It's already a generator:

```
In [32]: %time sum(range(200000000))
CPU times: user 3.86 s, sys: 9.2 ms, total: 3.86 s
Wall time: 3.86 s
```

```
In [34]: %timeit any(map(lambda x: x > 2, range(200000000)))
940 ns ± 6.97 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

File: sklearn/model_selection/_search.py (hunk 1 above)
Comment (ryan-deak-zefr) [reply sklearn/model_selection/_search.py:844]: I'm not sure if this is possible.  The reason I say this is that we can't necessarily just take the data set size because I don't think that the cross validation iterator (the `cv` parameter in `GridSearchCV`) must iterate over the entire dataset.  So it should be based on the sizes defined by `cv`.

**Review by vecchp — APPROVED** [review]

File: sklearn/model_selection/_search.py (hunk 1 above)
Comment (ryan-deak-zefr) [reply sklearn/model_selection/_search.py:844]: Should I put a comment to this effect in the code?

File: sklearn/model_selection/_search.py (hunk 4 above)
Comment (zexuan-zhou) [reply sklearn/model_selection/_search.py:891]: I agree. But seems like putting correlation structure here is too complicated to implement

File: sklearn/model_selection/tests/test_validation.py [hunk 8]
```diff
@@ -1647,3 +1652,109 @@ def test_sample_weight():
...
+    (accuracy_score, True, False, [1, 999999, 1, 999999]),
+    (accuracy_score, True, False, [100000, 200000, 100000, 200000]),
+    (accuracy_score, True, False, [100000, 100000, 100000, 100000]),
+    (accuracy_score, True, False, [200000, 100000, 200000, 100000]),
+    (accuracy_score, True, False, [999999, 1, 999999, 1]),
+    (accuracy_score, True, False, [2000000, 1000000, 1, 999999]),
+    (precision_score, True, False, [2000000, 1000000, 1, 999999]),
+    (log_loss, False, True, [2500000, 500000, 200000, 100000]),
+    (brier_score_loss, False, True, [2500000, 500000, 200000, 100000]),
+])
```
Comment (zexuan-zhou) [inline sklearn/model_selection/tests/test_validation.py:1667]: Should we test more scorers?

Comment (ryan-deak-zefr) [reply sklearn/model_selection/tests/test_validation.py:1667]: I don't think it's necessary.  We could, but this is way better than what seems to currently exist in sklearn.  I also have https://gist.github.com/deaktator/94545f807f139eba2c8a15381f2495e0 for which I'd like to submit a future PR to sklearn.  It's just test code.  Not any change to implementation.
